s3torchconnector/src/s3torchconnector/s3reader/constructor.py

```diff
@@ -1,12 +1,71 @@
 #  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 #  // SPDX-License-Identifier: BSD
 
+import logging
 from functools import partial
-from typing import Optional
-
-from .protocol import S3ReaderConstructorProtocol
+from typing import TYPE_CHECKING, Optional, List, Dict, Union
+from collections import defaultdict
+
+from .s3reader import S3Reader
+from .protocol import (
+    S3ReaderConstructorProtocol,
+    DCPS3ReaderConstructorProtocol,
+)
 from .sequential import SequentialS3Reader
 from .ranged import RangedS3Reader
+from .dcp_optimized import DCPOptimizedS3Reader, ItemRange, DEFAULT_MAX_GAP_SIZE
+
+if TYPE_CHECKING:
+    from torch.distributed.checkpoint.planner import ReadItem
+    from torch.distributed.checkpoint.metadata import MetadataIndex
+    from torch.distributed.checkpoint.filesystem import _StorageInfo
+
+log = logging.getLogger(__name__)
+
+
+class DCPOptimizedConstructor:
+    def __init__(self, max_gap_size: Union[int, float] = DEFAULT_MAX_GAP_SIZE) -> None:
+
+        if max_gap_size < 0:
+            raise ValueError("max_gap_size must be non-negative")
+
+        self._item_ranges_by_file: Dict[str, List[ItemRange]] = {}
+        self._max_gap_size: Union[int, float] = max_gap_size
+
+    def set_item_ranges_by_file(
+        self,
+        plan_items: "List[ReadItem]",
+        storage_data: "Dict[MetadataIndex, _StorageInfo]",
+    ) -> None:
+
+        if not plan_items:
+            return  # Allow lack of plan_items, for SequentialS3Reader fallbacks
```

Comment:
Is this correct? Did we decide what we wanted to do in case this method was called multiple times?